homeassistant/components/homekit/__init__.py

```diff
@@ -510,32 +526,40 @@ def setup(self, async_zeroconf_instance, uuid):
         if os.path.exists(persist_file):
             self.driver.load()
 
-    async def async_reset_accessories(self, entity_ids):
+    async def async_reset_accessories(self, entity_ids: Iterable[str]) -> None:
         """Reset the accessory to load the latest configuration."""
         if not self.bridge:
             await self.async_reset_accessories_in_accessory_mode(entity_ids)
             return
         await self.async_reset_accessories_in_bridge_mode(entity_ids)
 
-    async def async_reset_accessories_in_accessory_mode(self, entity_ids):
+    async def async_reset_accessories_in_accessory_mode(
+        self, entity_ids: Iterable[str]
+    ) -> None:
         """Reset accessories in accessory mode."""
-        acc = self.driver.accessory
+        assert self.driver is not None
+        acc = cast(HomeAccessory, self.driver.accessory)
         if acc.entity_id not in entity_ids:
             return
         await acc.stop()
         if not (state := self.hass.states.get(acc.entity_id)):
             _LOGGER.warning(
-                "The underlying entity %s disappeared during reset", acc.entity
+                "The underlying entity %s disappeared during reset", acc.entity_id
             )
             return
         if new_acc := self._async_create_single_accessory([state]):
             self.driver.accessory = new_acc
             self.hass.async_add_job(new_acc.run)
             await self.async_config_changed()
 
-    async def async_reset_accessories_in_bridge_mode(self, entity_ids):
+    async def async_reset_accessories_in_bridge_mode(
+        self, entity_ids: Iterable[str]
+    ) -> None:
         """Reset accessories in bridge mode."""
+        assert self.aid_storage is not None
+        assert self.bridge is not None
```

Comment:
`aid_storage` could be injected with minimal refactoring since we have direct control over this one.

`bridge` cannot since it may not be a bridge (accessory mode).

`HomeDriver` injection is possible with a major refactor of the integration and tests (1000s of lines). Its also based on upstream `AccessoryDriver` which needs to be created in the executor since it does I/O (https://github.com/ikalchev/HAP-python/blob/dev/pyhap/accessory_driver.py#L287) when its created and likely can't be changed without a major refactor upstream and breaking changes to other projects using `HAP-python`

Reply:
Ah right, no ideal, makes sense.

For the driver/aid_storage this approach might work too:

```diff
diff --git a/homeassistant/components/homekit/__init__.py b/homeassistant/components/homekit/__init__.py
index e2c442b8f4..b71fcadce3 100644
--- a/homeassistant/components/homekit/__init__.py
+++ b/homeassistant/components/homekit/__init__.py
@@ -465,6 +465,8 @@ def _async_register_events_and_services(hass: HomeAssistant) -> None:
 class HomeKit:
     """Class to handle all actions between HomeKit and Home Assistant."""
 
+    driver: HomeDriver
+
     def __init__(
         self,
         hass: HomeAssistant,
@@ -500,7 +502,6 @@ class HomeKit:
         self.status = STATUS_READY
 
         self.bridge: HomeBridge | None = None
-        self.driver: HomeDriver | None = None
 
     def setup(self, async_zeroconf_instance: AsyncZeroconf, uuid: UUID) -> None:
         """Set up bridge and accessory driver."""
@@ -537,7 +538,6 @@ class HomeKit:
         self, entity_ids: Iterable[str]
     ) -> None:
         """Reset accessories in accessory mode."""
-        assert self.driver is not None
         acc = cast(HomeAccessory, self.driver.accessory)
         if acc.entity_id not in entity_ids:
             return
@@ -591,7 +591,6 @@ class HomeKit:
 
     async def async_config_changed(self) -> None:
         """Call config changed which writes out the new config to disk."""
-        assert self.driver is not None
         await self.hass.async_add_executor_job(self.driver.config_changed)
 
     def add_bridge_accessory(self, state: State) -> HomeAccessory | None:
@@ -612,7 +611,6 @@ class HomeKit:
             )
 
         assert self.aid_storage is not None
-        assert self.driver is not None
         assert self.bridge is not None
         aid = self.aid_storage.get_or_allocate_aid_for_entity_id(state.entity_id)
         conf = self._config.get(state.entity_id, {}).copy()
@@ -729,7 +727,6 @@ class HomeKit:
             return
         self._async_register_bridge()
         _LOGGER.debug("Driver start for %s", self._name)
-        assert self.driver is not None
         await self.driver.async_start()
         async with self.hass.data[DOMAIN][PERSIST_LOCK]:
             await self.hass.async_add_executor_job(self.driver.persist)
@@ -742,7 +739,6 @@ class HomeKit:
     @callback
     def _async_show_setup_message(self) -> None:
         """Show the pairing setup message."""
-        assert self.driver is not None
         async_show_setup_message(
             self.hass,
             self._entry_id,
@@ -754,7 +750,6 @@ class HomeKit:
     @callback
     def async_unpair(self) -> None:
         """Remove all pairings for an accessory so it can be repaired."""
-        assert self.driver is not None
         state = self.driver.state
         for client_uuid in list(state.paired_clients):
             # We need to check again since removing a single client
@@ -770,7 +765,6 @@ class HomeKit:
     @callback
     def _async_register_bridge(self) -> None:
         """Register the bridge as a device so homekit_controller and exclude it from discovery."""
-        assert self.driver is not None
         dev_reg = device_registry.async_get(self.hass)
         formatted_mac = device_registry.format_mac(self.driver.state.mac)
         # Connections and identifiers are both used here.
@@ -826,7 +820,6 @@ class HomeKit:
         self, entity_states: list[State]
     ) -> HomeAccessory | None:
         """Create a single HomeKit accessory (accessory mode)."""
-        assert self.driver is not None
         if not entity_states:
             _LOGGER.error(
                 "HomeKit %s cannot startup: entity not available: %s",
@@ -849,7 +842,6 @@ class HomeKit:
         self, entity_states: Iterable[State]
     ) -> HomeAccessory:
         """Create a HomeKit bridge with accessories. (bridge mode)."""
-        assert self.driver is not None
         self.bridge = HomeBridge(self.hass, self.driver, self._name)
         for state in entity_states:
             self.add_bridge_accessory(state)
@@ -886,7 +878,6 @@ class HomeKit:
 
         if acc is None:
             return False
-        assert self.driver is not None
         # No need to load/persist as we do it in setup
         self.driver.accessory = acc
         return True
@@ -897,7 +888,6 @@ class HomeKit:
             return
         self.status = STATUS_STOPPED
         _LOGGER.debug("Driver stop for %s", self._name)
-        assert self.driver is not None
         await self.driver.async_stop()
 
     @callback
```

It does, however, make some tests fail (which is kinda a good thing I guess?)

Reply:
I think splitting the class into something like HomeKitBridge and HomeKitAccessory and making a base class is a good idea.  It is difficult to follow which is the accessory code and which is the bridge code... something for a future PR though.

Reply:
I tried injecting `aidstorage` but it resulted in 38 tests needing to be refactored so I didn't commit it.